### ISSUE TITLE
needless_range_loop false positive

### ISSUE THREAD

**Issue by dylanede** [Manishearth/rust-clippy#1879]
I have code like this:

```rust
for i in 1..vector.len() {
    if f(&vector[i - 1], &vector[i]) {
        g(&mut vector, i - 1, i);
    }
}
```

This is explicitly in this form as the interior of the loop requires mutable access to `vector`. I cannot change this into a loop using an iterator on `vector`, as the lint suggests.
